### PR TITLE
Sprint2

### PR DESCRIPTION
**Pull request by DeanBullockCGI**
Sprint 2 Changes
Has migrations

CF-25	Link to Formfinder for Complaints	"As a citizen I want to be able to quickly find forms that I need. As a citizen I want to be able to enter a complaint.
As a Citizen I would like a see a link to Formfinder in C&TF so I can easily and quickly send complaints back to the HMCTS via the feedback tool.

CF-72	Housing Possession Area of Law Text Change	"As an Administrator I want to ensure the Areas of Law search text is appropriate to Housing Possessions. 

CF-71	Child area of law text change	"As an Administrator I want to ensure the Areas of Law search text is appropriate to Child cases. As per UX/UI Design. 

CF-70	Divorce - Law Area Text Change	"As an Administrator I want to ensure the Areas of Law search text is appropriate to Divorce cases.

CF-61	Contacting the CCMCC	"As a CCMCC telephonist I want the C&TF to clearly inform citizens when and how to call the CCMCC and when/how to call the court. The CCMCC take 500 'failure' calls a month because it is unclear to citizens who and when to call.

CF-19	Improving Work Types	"Cases Heard Title on Court Page and search results (replacing areas of law). 
Text change 'Cases Heard at this Venue'

CF-31	Linking info when a court is closed	"As a citizen I want to know how to find an alternative court if the one I used previously has been closed. 

CF-28	Inclusion of non-HMCTS Buildings	"As a citizen I need to know if there are other venues that may be used for my tribunal if I have an immigration, Social Security or Employment case.

CF-59	Remove Maps in Search Results	"As a Citizen I want to see a map link on the relevant court page that I can click on, rather than be confused by having one in the search results page that does not work.

### CODE REVIEW

File: courtfinder/search/tests/test_courtcode.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import json
+from django.test import TestCase, Client
+from search.models import *
```
Comment (lgarvey) [inline courtfinder/search/tests/test_courtcode.py]: wildcard imports aren't good - can you specify the objects you're importing?

File: courtfinder/search/tests/test_courtcode.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+from django.conf import settings
```
Comment (lgarvey) [inline courtfinder/search/tests/test_courtcode.py]: Typically we order imports like this:

core python 
library (e.g. django)
application

It's helpful to enforce some consistency in terms of improving the codebase.  

import json

from django.test import TestCase, Client
from django.conf import settings

from search.ingest import Ingest
from search.models import X, Y, Z

etc.

File: courtfinder/courts/views.py [hunk 3]
```diff
@@ -72,7 +72,8 @@ def format_court(court):
                   'contacts': collapse(contacts, 'name', 'numbers'),
                   'directions': court.directions if court.directions else None,
                   'alert': court.alert if court.alert and court.alert.strip() != '' else None,
-                  'parking': court.parking if court.parking else None}
+                  'parking': court.parking if court.parking else None,
```
Comment (cgiHarland) [inline courtfinder/courts/views.py]: It scans better as well, so have changed here but obviously not changed all the others.

File: courtfinder/search/court_search.py [hunk 4]
```diff
@@ -60,7 +61,14 @@ def __init__( self, postcode=None, area_of_law=None, single_point_of_entry=False
 
     def get_courts( self ):
         if hasattr(self, 'query'):
-            return self.__address_search(self.query)
+            numberRE = re.compile('^[0-9]{3,4}$')
+            courtNumber = numberRE.match(self.query)
+
+            if self.courtcode_search == True: #courtNumber:
```
Comment (cgiHarland) [inline courtfinder/search/court_search.py]: Changed as per comment and removed commented out statement.

File: courtfinder/search/court_search.py [hunk 5]
```diff
@@ -190,6 +198,23 @@ def __address_search( self, query ):
...
+
+        word_separator = re.compile(r'[^\w]+', re.UNICODE)
+        query_regex = ''.join(map(lambda word: "(?=.*\y"+word+"\y)", re.split(word_separator, query)))
+
+        court_number_results =  sorted(Court.objects.filter(number__iregex=query_regex))
+
+        # put it all together and remove duplicates
+        results = list(OrderedDict.fromkeys(chain(court_number_results)))
+
+        return [result for result in results if result.displayed]
```
Comment (cgiHarland) [inline courtfinder/search/court_search.py]: Applied.

File: courtfinder/search/court_search.py [hunk 6]
```diff
@@ -190,6 +198,23 @@ def __address_search( self, query ):
...
+        Retrieve court(s) by court code, order and remove duplicates
+        """
+
+        # First we get courts whose court code contains the query string
+        # (for these courts sorted to show the courts with the highest number of areas of law first)
+
+        word_separator = re.compile(r'[^\w]+', re.UNICODE)
+        query_regex = ''.join(map(lambda word: "(?=.*\y"+word+"\y)", re.split(word_separator, query)))
+
+        court_number_results =  sorted(Court.objects.filter(number__iregex=query_regex))
```
Comment (cgiHarland) [inline courtfinder/search/court_search.py]: I did do this before but removed it, as it was not as the other code was working. Have replaced with order_by('name').

File: courtfinder/search/tests/test_search.py [hunk 7]
```diff
@@ -317,3 +426,59 @@ def test_data_status(self):
...
+    def test_for_no_venue_links(self):
+        c = Client()
+        response = c.get('/search/results?aol=Bankruptcy&postcode=SE15')
+        self.assertEqual(200, response.status_code)
+        self.assertNotIn('https://www.gov.uk/guidance/employment-tribunal-offices-and-venues', response.content)
+        self.assertNotIn('http://sscs.venues.tribunals.gov.uk/venues/venues.htm', response.content)
+        
+    def test_gov_uk_links_exist(self):
+        
+        try:
```
Comment (cgiHarland) [inline courtfinder/search/tests/test_search.py]: I've removed the other test to make things clearer.

File: courtfinder/search/views.py [hunk 8]
```diff
@@ -34,11 +35,25 @@
...
+    url = url + searchby
+    return HttpResponseRedirect(reverse(url))
+
+
 def aol(request):
     areas_of_law = AreaOfLaw.objects.all().exclude(name='High court').order_by('name')
     aol = request.GET.get('aol', 'All')
     for area in areas_of_law:
-        area.description = areas_of_law_description[area.name]
+        try:
```
Comment (cgiHarland) [inline courtfinder/search/views.py]: Applied.

File: courtfinder/search/tests/test_courtcode.py (hunk 1 above)
Comment (cgiHarland) [inline courtfinder/search/tests/test_courtcode.py]: Applied.

File: courtfinder/search/tests/test_courtcode.py (hunk 2 above)
Comment (cgiHarland) [inline courtfinder/search/tests/test_courtcode.py]: Applied.

**Pull request merged**